src/physics/adv_fluxcorr.f90

```diff
@@ -0,0 +1,193 @@
+        real, dimension(its-1:ite+2,  kms:kme,jts-1:jte+1),  intent(in) :: u
+        real, dimension(its-1:ite+1,  kms:kme,jts-1:jte+2),  intent(in) :: v
+        
+        real, dimension(its-1:ite+2,kms:kme,  jts-1:jte+1),intent(inout)          :: flux_x
+        real, dimension(its-1:ite+1,kms:kme,  jts-1:jte+2),intent(inout)          :: flux_y
+        real, dimension(its-1:ite+1,kms:kme+1,jts-1:jte+1),intent(inout)    :: flux_z
+        
+        
+        real, dimension(its-1:ite+2,kms:kme,  jts-1:jte+1)         :: upwind_flux_x
+        real, dimension(its-1:ite+1,kms:kme,  jts-1:jte+2)         :: upwind_flux_y
+        real, dimension(its-1:ite+1,kms:kme+1,jts-1:jte+1)         :: upwind_flux_z
+        
+        real, dimension(its-1:ite+1,  kms:kme,jts-1:jte+1)   :: flux_in, scale_in, flux_out, scale_out, temp, qmax, qmin
+
+        !Initialize some internal variables
+        scale_in = 1.0
+        scale_out = 1.0
+              
+        qmax = q(its-1:ite+1,:,jts-1:jte+1)
+        qmin = q(its-1:ite+1,:,jts-1:jte+1)
+        
+        ! Get upwind fluxes
+        call upwind_flux3(q,u,v,w,upwind_flux_x,upwind_flux_z,upwind_flux_y,jaco,dz,rho)
+
+        !Compute concentration if upwind only was used
+        temp  = q(its-1:ite+1,:,jts-1:jte+1) - ((upwind_flux_x(its:ite+2,:,:) - upwind_flux_x(its-1:ite+1,:,:)) + &
+                                       (upwind_flux_y(:,:,jts:jte+2) - upwind_flux_y(:,:,jts-1:jte+1))) &
+                                   / (jaco*rho)                      
+        temp = temp - (upwind_flux_z(:,kms+1:kme+1,:) - upwind_flux_z(:,kms:kme,:)) / (dz*jaco*rho)
+        
+        ! We are now done with the input fluxes, so they can be overwritten to save memory
+        ! flux now represents the merged fluxes
+        flux_x = flux_x - upwind_flux_x
+        flux_y = flux_y - upwind_flux_y
+        flux_z = flux_z - upwind_flux_z
+        
+        ! Next compute max and min possible fluxes
+        flux_in = -((min(0.,flux_x(its:ite+2,:,:)) - max(0.,flux_x(its-1:ite+1,:,:))) + &
+                        (min(0.,flux_y(:,:,jts:jte+2)) - max(0.,flux_y(:,:,jts-1:jte+1)))) &
+                                   / (jaco*rho)                         
+        flux_in = flux_in - (min(0.,flux_z(:,kms+1:kme+1,:)) - max(0.,flux_z(:,kms:kme,:))) / (dz*jaco*rho)
+        
+        flux_out = ((max(0.,flux_x(its:ite+2,:,:)) - min(0.,flux_x(its-1:ite+1,:,:))) + &
+                        (max(0.,flux_y(:,:,jts:jte+2)) - min(0.,flux_y(:,:,jts-1:jte+1)))) &
+                                   / (jaco*rho)                         
+        flux_out = flux_out + (max(0.,flux_z(:,kms+1:kme+1,:)) - min(0.,flux_z(:,kms:kme,:))) / (dz*jaco*rho)
+        
+                            
+        where (u(its-1:ite+1,:,:) > 0)
+            qmax(:,:,:) = max(q(its-2:ite,:,jts-1:jte+1),qmax(:,:,:))
+            qmin(:,:,:) = min(q(its-2:ite,:,jts-1:jte+1),qmin(:,:,:))
+        else where (u(its:ite+2,:,:) < 0)
```

Comment:
what should happen for grid cells in which either
`u(i) > 0 and u(i+1)<0`
or
`u(i) < 0 and u(i+1)>0`

Reply:
shouldn't both qmax/min values be tested?  with the else where, only the first will be used to compute the max/min
(maybe the second case `u(i)<0` doesn't matter as it doesn't change the limit?)

Reply:
Good catch -- I see your point now. I'll change these else wheres to wheres. As for the second condition you outlined, yes, the flux correction works by identifying the minimum and maximum value around a cell and saying that we should not be creating any new mins and maxes as a product of our advection step (the monotonic condition). So if there are no fluxes from a given direction into a cell, we don't have to consider the concentration of that upwind cell in the min/max.